### ISSUE TITLE
Validation -- F1 Score, penalization

### ISSUE THREAD

**Issue by ad1m** [pulearn/pulearn#12]
I have been experimenting with a dataset and a few different classifiers with pulearn. I noticed that all classifiers I had trained had nice f1 scores but upon further inspection of one particular classifier (adaboost) I noticed that it predicted all samples as the positive class. Since the validation is done on the heldout positive samples, the F1 score can still remain high when in fact this classifier is not working. This made me think of a few questions: 

1. What are the standard methodologies used to evaluate a pulearner? Is it holding out a set of positives and computing f1, etc? Or is there any alternative standard methods?

2. Is there any standard for penalizing a pulclassifier that predicts all samples as one class. What are some strategies to make sure we do not simple predict one class?

**Comment by shaypal5** [comment]
Hey. Short on time right now.
Just popped to say that these are VERY good questions, and I stumbled upon them myself when using this approach in a couple of projects.

I found at least one paper which detailed its evaluation schema, and I also remember a place where it was formalized which metrics can be computed and which not. If I remember correctly (and this also makes sense intuitively), in classic PU learning you can compute recall but not precision, because you have no Negative labels, so any predicted positive might be true. Put otherwise, you can never measure the amount of False Positive, and as a result, you can't compute F1 score and anything else that depends on FPs.

BUT, I did find a paper that outlined how to compute a metric that IIRC is correlated with F1 (or it might be a lower bound). Which is cool.

I'm leaving this unread in my mailbox and I'll get back to this in the near future.

Thank you for asking! :)

**Comment by ydennisy** [comment]
hey @shaypal5 wondering if you have had time to work in this question?

I am also struggling in finding the "correct" way to try and generalise a metric in a PU learning scenario.

**Comment by shaypal5** [comment]
Not really, but slides 33 to 37 in a presentation I once prepared for an intro to PU Learning discusses metrics:

https://docs.google.com/presentation/d/1reAifZKvSy0FlxuOYiAksjYeAoLzn1wdI2d21LQ-COk/edit?usp=sharing

Also, slide 71 has references, and I believe the formalization I referenced can be found in one of the two articles referenced there.

Let me know what you find out!

**Comment by shaypal5** [comment]
I've since added slide 38, with specific references to papers dealing with metrics for PU Learning. This includes both Li & Liu 2003, which formalized the F1-like metric for PU, and a couple of much later papers (2015, 2019) on estimating standard metrics for the case of PU (with upper and lower bounds and such):

https://docs.google.com/presentation/d/1reAifZKvSy0FlxuOYiAksjYeAoLzn1wdI2d21LQ-COk/edit?usp=sharing

**Comment by shaypal5** [comment]
This was partly addressed in PR #19 !
